Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py

```diff
@@ -1032,7 +1032,47 @@ def checkout(self, argument, prune=None):
                 cwd=self.root_path,
             ).returncode else self.commit()
 
+        match = self.dev_branches.match(argument)
         branch_remote = self.remote_for(argument)
+
+        # if branch is not a dev branch and has a remote
+        if not match and branch_remote:
+            path = '{}/{}'.format(branch_remote, argument)
+            local_head = self.commit(branch=argument)
+            remote_head = self.commit(branch=path)
+            branches_for_local = self.branches_for(local_head.hash)
```

Comment:
I don't think this quite gives us the answer we need. We only want to prompt the user if the user is at risk of losing data. The only way the user is at risk of losing data is if the ref they're targeting includes a change which doesn't exist on any remote. At first glance, this seems like a challenging problem to efficiently solve, I though it was `O(n)` where `n` is the number of remote branches we track (which in WebKit, may be in the low thousands).

However, @gsnedders showed me `git merge-base` a few months back, and that turns out to basically magically answer our question here.

If you invoke a command like this:
```
git merge-base <hash> remotes/<remote>/<branch-a> remotes/<remote>/<branch-b>...
```
you will end up with a hash. If that hash is equal to the `<hash>` you passed in as the first argument, then the provided hash is a member of at least one of the provided remote branches. Otherwise, you know that <hash> represents a commit which only exists locally.

I think this is the approach you want to take here. `git merge-base` runs extremely quickly , your only catch is that Windows has a limit of 32K characters in a single command, so you want to limit the number of arguments to your shell invocation. `git-webkit branch` (which uses `git merge-base` to achieve a similar goal) just caps the number of arguments at 512, which seems like a prudent approach here as well.

Reply:
I lack some of the context here, given I don't know when we're force pushing to the tracked remote.

Do we expect the former head of the tracked remote to continue to be on some other branch, because if not we could potentially end up with a very large number of orphaned commits which probably are of no interest to the author?

When it comes to `git merge-branch`, we should probably refactor the code that's currently in `program.branch.Branch.branch_point` to be somewhere else (in https://github.com/WebKit/WebKit/pull/13934#discussion_r1203069779 I seemingly suggested moving it to `Repository`, so maybe let's do that?).

That said, I don't know how common this workflow is, nor how much we should care about this.

Reply:
I think my previous implementation did give the right behaviour but your approach is more efficient, so I've used `branch_point`. I decided to keep the checks for what branches are associated with the commit, as I don't want to reset the branch if it isn't necessary. So, if there's a local change but the local branch is tracking the correct remote branch, a normal checkout is performed.

Reply:
#18445